common/src/main/java/io/crate/interval/PGIntervalParser.java

```diff
@@ -0,0 +1,167 @@
+
+                    int endMinutes = token.indexOf(':', endHours + 1);
+                    seconds = parseInteger(token.substring(endMinutes + 1));
+                    milliSeconds = parseMilliSeconds(token.substring(endMinutes + 1));
+
+                    if (offset == 1) {
+                        hours = -hours;
+                        minutes = -minutes;
+                        seconds = -seconds;
+                        milliSeconds = -milliSeconds;
+                    }
+                    valueToken = null;
+                } else {
+                    // This handles years, months, days for both, ISO and
+                    // Non-ISO intervals. Hours, minutes, seconds and microseconds
+                    // are handled for Non-ISO intervals here.
+                    if (token.startsWith("year")) {
+                        years = nullSafeIntGet(valueToken);
+                    } else if (token.startsWith("mon")) {
+                        months = nullSafeIntGet(valueToken);
+                    } else if (token.startsWith("day")) {
+                        days = nullSafeIntGet(valueToken);
+                    } else if (token.startsWith("week")) {
+                        days = nullSafeIntGet(valueToken) * 7;
+                    } else if (token.startsWith("hour")) {
+                        hours = nullSafeIntGet(valueToken);
+                    } else if (token.startsWith("min")) {
+                        minutes = nullSafeIntGet(valueToken);
+                    } else if (token.startsWith("sec")) {
+                        seconds = parseInteger(valueToken);
+                        milliSeconds = parseMilliSeconds(valueToken);
+                    }
+                }
+            }
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("Invalid interval format " + value);
+        }
+
+        if (years == null && months == null && days == null && hours == null && minutes == null && seconds == null) {
+            throw new IllegalArgumentException("Invalid interval format " + value);
+        }
+
+        Period period = buildPeriod(years, months, days, hours, minutes, seconds, milliSeconds);
+
+        if (!ISOFormat && value.endsWith("ago")) {
+            // Inverse the leading sign
+            period = period.negated();
+        }
+        return period;
+    }
+
+    private static int parseInteger(String value) {
+        return new BigDecimal(value).intValue();
+    }
+
+    private static Period buildPeriod(Integer years,
+                              Integer months,
+                              Integer days,
+                              Integer hours,
+                              Integer minutes,
+                              Integer seconds,
+                              Integer millis) {
+        Period period = new Period();
+
+        if (years != null) {
+            period = period.withYears(years);
```

Comment:
How is this now different from setting the values to 0? If we use `0` as default for the values we could probably also use the `new Period(year, month...)` constructor to avoid repeated re-creation of a Period instance.

Reply:
Good point, thank you!